File: Data/callingBedrockKB.js
//npm install @aws-sdk/client-bedrock-agent-runtime

const { BedrockAgentRuntimeClient, RetrieveAndGenerateCommand } = require("@aws-sdk/client-bedrock-agent-runtime");

/**
 * Query a Bedrock Knowledge Base
 * @param {string} knowledgeBaseId - The ID of the Bedrock Knowledge Base
 * @param {string} prompt - The user's query/prompt
 * @param {string} region - AWS region (default: us-east-1)
 * @returns {Promise} - Response containing the generated text and citations
 */
async function queryBedrockKnowledgeBase(knowledgeBaseId, prompt, region = "us-east-1") {
    // Initialize the Bedrock Agent Runtime client
    const client = new BedrockAgentRuntimeClient({ region });

    // Prepare the command
    const command = new RetrieveAndGenerateCommand({
        input: {
            text: prompt
        },
        retrieveAndGenerateConfiguration: {
            type: "KNOWLEDGE_BASE",
            knowledgeBaseConfiguration: {
                knowledgeBaseId: knowledgeBaseId,
                modelArn: `arn:aws:bedrock:${region}::foundation-model/anthropic.claude-3-sonnet-20240229-v1:0`
            }
        }
    });

    try {
        // Send the command
        const response = await client.send(command);
        
        // Extract the generated text
        const generatedText = response.output.text;
        
        // Extract citations if available
        const citations = response.citations || [];
        
        return {
            answer: generatedText,
            citations: citations,
            sessionId: response.sessionId
        };
    } catch (error) {
        console.error("Error querying Bedrock Knowledge Base:", error);
        throw error;
    }
}

// Example usage
async function main() {
    const knowledgeBaseId = "YOUR_KB_ID_HERE";
    const prompt = "What is Amazon Bedrock?";
    
    try {
        const result = await queryBedrockKnowledgeBase(knowledgeBaseId, prompt);
        console.log("Answer:", result.answer);
        console.log("\nCitations:", JSON.stringify(result.citations, null, 2));
    } catch (error) {
        console.error("Failed to query knowledge base:", error);
    }
}

// Uncomment to run
// main();

module.exports = { queryBedrockKnowledgeBase };

File: Data/calling_bedrock_kb.py
# pip install boto3
import boto3
from typing import Dict, List, Optional
import json

def query_bedrock_knowledge_base(
    knowledge_base_id: str,
    prompt: str,
    region: str = "us-east-1",
    model_arn: Optional[str] = None
) -> Dict:
    """
    Query a Bedrock Knowledge Base
    
    Args:
        knowledge_base_id (str): The ID of the Bedrock Knowledge Base
        prompt (str): The user's query/prompt
        region (str): AWS region (default: us-east-1)
        model_arn (str): Optional model ARN. If not provided, uses Claude 3 Sonnet
        
    Returns:
        dict: Response containing the generated text and citations
    """
    # Initialize the Bedrock Agent Runtime client
    client = boto3.client(
        service_name='bedrock-agent-runtime',
        region_name=region
    )
    
    # Set default model if not provided
    if model_arn is None:
        model_arn = f"arn:aws:bedrock:{region}::foundation-model/anthropic.claude-3-sonnet-20240229-v1:0"
    
    try:
        # Call the retrieve_and_generate API
        response = client.retrieve_and_generate(
            input={
                'text': prompt
            },
            retrieveAndGenerateConfiguration={
                'type': 'KNOWLEDGE_BASE',
                'knowledgeBaseConfiguration': {
                    'knowledgeBaseId': knowledge_base_id,
                    'modelArn': model_arn
                }
            }
        )
        
        # Extract the generated text
        generated_text = response['output']['text']
        
        # Extract citations if available
        citations = response.get('citations', [])
        
        return {
            'answer': generated_text,
            'citations': citations,
            'session_id': response.get('sessionId')
        }
        
    except Exception as error:
        print(f"Error querying Bedrock Knowledge Base: {error}")
        raise


# Example usage
def main():
    knowledge_base_id = "YOUR_KB_ID_HERE"
    prompt = "What is Amazon Bedrock?"
    
    try:
        result = query_bedrock_knowledge_base(knowledge_base_id, prompt)
        print("Answer:", result['answer'])
        print("\nCitations:", json.dumps(result['citations'], indent=2))
    except Exception as e:
        print(f"Failed to query knowledge base: {e}")


if __name__ == "__main__":
    main()
